dali/pipeline/data/tensor_vector_test.cc

```diff
@@ -52,6 +59,633 @@ constexpr cudaStream_t cuda_stream = 0;
+      t.set_type(type);
+    }},
+    {"layout", [layout](TensorVector<Backend> &t) {
+      t.SetLayout(layout);
+    }}
+  };
+}
+
+
+TYPED_TEST(TensorVectorSuite, PartialSetupCopyMultiGPU) {
+  int ndev = 0;
+  CUDA_CALL(cudaGetDeviceCount(&ndev));
+  if (ndev < 2)
+    GTEST_SKIP() << "This test requires at least 2 CUDA devices";
+  constexpr bool is_device = std::is_same_v<TypeParam, GPUBackend>;
+  constexpr int device = 1;
+  const auto order = is_device ? AccessOrder(cuda_stream, device) : AccessOrder::host();
+  Tensor<TypeParam> t;
+  t.set_device_id(device);
+  t.set_order(order);
+  t.set_pinned(true);
+  t.Resize({3, 4, 5}, DALI_INT32);
+  t.SetLayout("HWC");
+
+  // set size to be checked. Copy doesn't make sense
+  auto setups = CopyRequiredSetters<TypeParam>(3, DALI_INT32, "HWC");
+  for (size_t excluded = 0; excluded < setups.size(); excluded++) {
+    std::vector<size_t> idxs(setups.size());
+    std::iota(idxs.begin(), idxs.end(), 0);
+    do {
+      TensorVector<TypeParam> tv;
+      tv.SetContiguous(BatchState::Noncontiguous);
+      tv.SetSize(4);
+      for (auto idx : idxs) {
+        if (idx == excluded) {
+          continue;
+        }
+        setups[idx].second(tv);
+      }
+      try {
+        tv.UnsafeCopySample(0, t);
+        FAIL() << "Exception was expected with excluded: " << setups[excluded].first;
+      } catch(std::runtime_error &e) {
+        auto expected = "Sample must have the same " + setups[excluded].first;
+        EXPECT_NE(std::string(e.what()).rfind(expected), std::string::npos)
+            << expected << "\n====\nvs\n====\n"
+            << e.what();
+      } catch (...) {
+        FAIL() << "Unexpected exception";
+      }
+    } while (std::next_permutation(idxs.begin(), idxs.end()));
+  }
+}
+
+
+TYPED_TEST(TensorVectorSuite, FullSetupSetMultiGPU) {
+  int ndev = 0;
+  CUDA_CALL(cudaGetDeviceCount(&ndev));
+  if (ndev < 2)
+    GTEST_SKIP() << "This test requires at least 2 CUDA devices";
+  constexpr bool is_device = std::is_same_v<TypeParam, GPUBackend>;
+  constexpr int device = 1;
+  const auto order = is_device ? AccessOrder(cuda_stream, device) : AccessOrder::host();
+  Tensor<TypeParam> t;
+  t.set_device_id(device);
+  t.set_order(order);
+  t.set_pinned(true);
+  t.Resize({3, 4, 5}, DALI_INT32);
+  t.SetLayout("HWC");
+
+  // set size to be checked. Copy doesn't make sense
+  auto setups = SetRequiredSetters<TypeParam>(3, DALI_INT32, "HWC", true, device);
+  std::vector<size_t> idxs(setups.size());
+  std::iota(idxs.begin(), idxs.end(), 0);
+  do {
+    TensorVector<TypeParam> tv;
+    tv.SetContiguous(BatchState::Noncontiguous);
+    tv.SetSize(4);
+    for (auto idx : idxs) {
+      setups[idx].second(tv);
+    }
+    tv.UnsafeSetSample(0, t);
+    EXPECT_EQ(tv[0].raw_data(), t.raw_data());
+  } while (std::next_permutation(idxs.begin(), idxs.end()));
+}
+
+
+namespace {
+
+void FillWithNumber(TensorVector<CPUBackend> &tv, uint8_t number) {
```

Comment:
why not a template like below?

Reply:
Because I didn't need it in that test. But I can extend.

Reply:
just found it odd that one functions is a template and this one is not. Up to you